test/cpp_api_parity/utils.py

```diff
@@ -0,0 +1,225 @@
+#
+# For example:
+# ```
+# def bceloss_weights_no_reduce_test():
+#     t = torch.randn(15, 10).gt(0).double()
+#     weights = torch.rand(10)
+#     return dict(
+#         fullname='BCELoss_weights_no_reduce',
+#         constructor=wrap_functional(
+#             lambda i: F.binary_cross_entropy(i, t.type_as(i),
+#                                              weight=weights.type_as(i), reduction='none')),
+#         cpp_function_call='F::binary_cross_entropy(i, t.to(i.options()), F::BinaryCrossEntropyFuncOptions().weight(weights.to(i.options())).reduction(torch::kNone))',
+#         input_fn=lambda: torch.rand(15, 10).clamp_(2.8e-2, 1 - 2.8e-2),
+#         cpp_arg_symbol_map={'i': 'input', 't': t, 'weights': weights},
+#         reference_fn=lambda i, p, m: -(t * i.log() + (1 - t) * (1 - i).log()) * weights,
+#     )
+# ```
+def compute_arg_dict(test_params_dict, test_instance):
+    arg_dict = {
+        'input': [],
+        'target': [],
+        'extra_args': [],
+        'other': [],
+    }
+
+    def put_args_into_arg_dict(arg_type, arg_type_prefix, args):
+        for i, arg in enumerate(args):
+            arg_dict[arg_type].append(CppArg(name=arg_type_prefix + str(i), value=arg))
+
+    put_args_into_arg_dict('input', 'i', convert_to_list(test_instance._get_input()))
+    if is_criterion_test(test_instance):
+        put_args_into_arg_dict('target', 't', convert_to_list(test_instance._get_target()))
+    if test_instance.extra_args:
+        put_args_into_arg_dict('extra_args', 'e', convert_to_list(test_instance.extra_args))
+
+    cpp_arg_symbol_map = test_params_dict.get('cpp_arg_symbol_map', {})
+    for arg_name, arg_value in cpp_arg_symbol_map.items():
+        if isinstance(arg_value, str):
+            if arg_value == 'input':
+                arg_dict['other'].append(CppArg(name=arg_name, value=test_instance._get_input()))
+            else:
+                raise RuntimeError("`{}` has unsupported string value: {}".format(arg_name, arg_value))
+        elif isinstance(arg_value, torch.Tensor):
+            arg_dict['other'].append(CppArg(name=arg_name, value=arg_value))
+        else:
+            raise RuntimeError("`{}` has unsupported value: {}".format(arg_name, arg_value))
+
+    return arg_dict
+
+def decorate_test_fn(test_fn, test_cpp_api_parity, test_cuda, has_impl_parity, device):
+    test_fn = unittest.skipIf(not test_cpp_api_parity, "Excluded from C++ API parity tests")(test_fn)
+
+    if device == 'cuda':
+        test_fn = unittest.skipIf(not TEST_CUDA, "CUDA unavailable")(test_fn)
+        test_fn = unittest.skipIf(not test_cuda, "Excluded from CUDA tests")(test_fn)
+
+    # If `Implementation Parity` entry in parity table for this module is `No`,
+    # or `has_parity` entry in test params dict is `False`, we mark the test as
+    # expected failure.
+    if not has_impl_parity:
+        test_fn = unittest.expectedFailure(test_fn)
+
+    return test_fn
+
+extra_msg_on_failure = ''
```

Comment:
What is this used for?

Reply:
I used to put `MESSAGE_HOW_TO_FIX_CPP_PARITY_TEST_FAILURE` in `test/test_cpp_api_parity.py` and then inject it here, but I think it's bad code structure and just moved `MESSAGE_HOW_TO_FIX_CPP_PARITY_TEST_FAILURE` to this file instead :D